src/fusion_algorithm.cpp

```diff
@@ -0,0 +1,325 @@
+
+    for (int k = 0; k < number_of_sensors * number_of_sensors; ++k) {
+        degree_matrix_temp[k] = degree_matrix[k];
+    }
+
+    gsl_matrix_view m = gsl_matrix_view_array(degree_matrix_temp,
+                                              number_of_sensors,
+                                              number_of_sensors);
+
+    gsl_vector *eval = gsl_vector_alloc(number_of_sensors);
+    gsl_matrix *evec = gsl_matrix_alloc(number_of_sensors, number_of_sensors);
+
+    gsl_eigen_symmv_workspace *w = gsl_eigen_symmv_alloc(number_of_sensors);
+
+    gsl_eigen_symmv(&m.matrix, eval, evec, w);
+
+    gsl_eigen_symmv_free(w);
+
+    gsl_eigen_symmv_sort(eval, evec, GSL_EIGEN_SORT_VAL_DESC);
+
+
+    for (int i = 0; i < number_of_sensors; i++) {
+        eigenvalues[i] = gsl_vector_get(eval, i);
+        gsl_vector_view evec_i = gsl_matrix_column(evec, i);
+
+        double *vec = (double *) calloc(number_of_sensors, sizeof(double));
+        printf("Eigenvalues : %f - ", eigenvalues[i]);
+        printf("Eigenvector : [");
+        for (int j = 0; j < number_of_sensors; ++j) {
+            vec[j] = evec_i.vector.data[j];
+            printf("%f, ", vec[j]);
+        }
+        printf("]\n");
+        eigenvectors[i] = vec;
+    }
+
+    free(degree_matrix_temp);
+    gsl_vector_free(eval);
+    gsl_matrix_free(evec);
+}
+
+
+double **get_principal_components(double *degree_matrix, double **eignvectors, int number_of_sensors) {
+
+    double **principal_components = (double **) calloc(number_of_sensors, sizeof(double *));
+    for (int i = 0; i < number_of_sensors; i++) {
+        principal_components[i] = (double *) calloc(number_of_sensors, sizeof(double));
+    }
+
+    gsl_matrix *p_eigen_vec = gsl_matrix_alloc(number_of_sensors, number_of_sensors);
+    gsl_matrix *p_support_degree = gsl_matrix_alloc(number_of_sensors, number_of_sensors);
+    gsl_matrix *p_y = gsl_matrix_alloc(number_of_sensors, number_of_sensors);
+
+    for (int i = 0; i < number_of_sensors; i++) {
+        for (int j = 0; j < number_of_sensors; j++) {
+            gsl_matrix_set(p_eigen_vec, i, j, eignvectors[i][j]);
+            gsl_matrix_set(p_support_degree, i, j,
+                           degree_matrix[j + (i * number_of_sensors)]);
+            gsl_matrix_set(p_y, i, j, 0.0);
+        }
+    }
+
+
+    gsl_blas_dgemm(CblasNoTrans, CblasNoTrans,
+                   1.0, p_eigen_vec, p_support_degree,
+                   0.0, p_y);
+
+    printf("\n\nPrincipal components : \n");
+    for (int rows = 0; rows < number_of_sensors; rows++) {
+
+        for (int cols = 0; cols < number_of_sensors; cols++) {
+            principal_components[rows][cols] = gsl_matrix_get(p_y, rows, cols);
+            printf("%f, ", principal_components[rows][cols]);
+        }
+    }
+    printf("\n");
+    return principal_components;
+}
+
+double *get_contribution_rates(double *eigenvalues, int number_of_sensors) {
```

Comment:
need to verify if the memory is freed once this all these allocations are done with